jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/ChordalNiceTreeDecomposition.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * (C) Copyright 2018-2018, by Ira Justus Fesefeldt and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.decomposition;
+
+import java.util.*;
+
+import org.jgrapht.*;
+import org.jgrapht.alg.cycle.*;
+
+/**
+ * A builder for a nice decomposition for chordal graphs. See {@link NiceTreeDecompositionBase} for
+ * an explanation of nice decomposition.
+ * <p>
+ * This builder uses the perfect elimination order from {@link ChordalityInspector} to iterate over
+ * the graph. For every node it generates a node for the predecessors of the current node according
+ * to the perfect elimination order and builds a path to such a node from the node where the
+ * greatest predecessor was introduced.
+ * <p>
+ * The complexity of this algorithm is in $\mathcal{O}(|V|(|V|+|E|))$.<br>
+ * Consider the every node in the nice tree decomposition: There are exactly $|V|-1$ many forget
+ * nodes. There are at most $2(|V|-1)$ additionally nodes because of a join nodes. Every join node
+ * creates one additional path from root to a leaf, every such path can contain for every vertex at
+ * most one introduce node, which yields $|V|^2$ introduce nodes. Now considering the bags of the
+ * introduce nodes. On a path from root to a leaf we have at most one introduce node for every
+ * introduced vertex. Since this introduced vertex is part of the clique of the least ancestor
+ * forget node, the corresponding bag of the introduce node is smaller than the set of neighbors of
+ * this vertex. Thus the time complexity is in $\mathcal{O}(|V|(|V|+|E|))$.
+ * <p>
```

Comment:
We generate `|V| - 1` forget nodes and build `|V|-1` paths to previously added nodes in the tree. There is at most one join node generated per such path, therefore there are at most `|V|-1` join nodes. At least we are calling addJoinNode once per vertex from the perfect elimination order. 

I don't quite understand why the running time of the algorithm is O(V(V + E)). Computing order predecessors and finding one with the highest index contributes O(V + E) overall. Creating and adding forget nodes run in O(V+ E) since the sum of the cardinalities of all predecessors sets equals to |E|. I'm uncertain about the running time complexity of deleting the nodes while building and finishing the paths. Every vertex can be O(deg(v)) times the higher predecessor, consequently it can be deleted O(deg(v)) times (deleting a node = adding "introduce" node). Deleting a node takes O(deg(v)) time since we a dealing with bags which contain cliques. This is done for every vertex, that's why the running time is (V*vertex_max_deg^2 + E) which is O(VE^2) worst case.  Can we have a better complexity analysis? 

Another point is that we have O(V^2) upper bound on introduce nodes, but adding one introduce node runs in O(vertex_max_deg) time because of set copying. So, the running time is at least O(V^2*vertex_max_deg). Can we aggregate this somehow?

Reply:
I do not quite get your first point. But for the second one, I do not agree. Adding one introduce node runs in O(|b(parent)|) - which is important, since in this case, the parent consits of a clique. Thus adding an introduce node for the vertex v costs in this case O(|Neighbour(v)|). Which is far less then O(vertex_max_deg). But I guess, I can try to do another analysis. Maybe this one is more elegant.

First lets proof the following lemma: every bag of a node in the tree decomposition consists of a clique. Proof by induction: the roots bag is obviously a (trivial) clique. Additionally, adding a forget node produces a node with a clique by construction, because the bag consists of the forgotten element and its predecessors. 
Adding join nodes and adding introduce nodes only adds nodes, which bags are a subset of their parents bag. Since every subset of a clique is a clique, their bags are a clique by induction hypothesis. 

Now there are at most |V| different pathes from the root to a leaf, because every vertex can create at most one new path. Consider any of them. For now we ignore the join nodes. Because the tree is a tree decomposition, every vertex v can be introduced and forgotten at most ones. Adding a forget node generates a node with a clique of v. Adding an introduce node, generates a node with a clique of v, whithout v (because v is in the bag of its parent and the parents bag is a clique). Thus the cost of adding a forget node or adding an introduce node for one v is in O(|Neighbour(v)|). Now every vertex has at most one introduce node and one forget node on this path, which adds up to O(|V|+|E|) for one path. Since there are at most |V| different paths, we have O(|V|(|V|+|E|)).

Now considering the join nodes: There are 2*(|V|-1) join nodes, everyone can be generated with at most vertex_max_deg sized bags, thus they add up to O(|V|*vertex_max_deg). Simmilar calculating the toIntroduce sets add up to O(|V|*vertex_max_deg). In total again O(|V|(|V|+|E|)).

Reply:
Now the analysis is clear for me, thanks! One more issue to resolve: why there are 2*(V-1) joins (at most)? We are adding one join node per forget node, consequently there are at most (V-1) join nodes.

Reply:
You are right. There are (V-1) join nodes, but adding join nodes generate 2 nodes. But these two nodes are not (necessary) join nodes according to the definition.

Reply:
All right, thanks! Let's use the following Javadoc:
```
/**
 * Builds a nice tree decomposition of chordal graphs. See {@link NiceTreeDecompositionBase} for
 * an explanation of the nice tree decomposition.
 * <p>
 * For a given perfect elimination order of the graph $G = (V, E)$ every vertex in $V(G)$ has an associated 
 * index in this order. The predecessors of a vertex $x \in V(G)$ are those vertices incident to $x$ that 
 * have smaller index than $x$ (a vertex can have no predecessors). This class uses a perfect elimination 
 * order produced by {@link ChordalityInspector} to iteratively add vertices of $G$ to some nodes in the 
 * tree $T$. Initially, it adds a bag to the decomposition with the first vertex from the perfect elimination 
 * order as a root of the tree. Then for all remaining vertices $v \in V(G)$ this class generates a node (bag) 
 * $s \in I(T)$ in the tree $T$ which contains vertex $v$ and all its predecessors. Then it builds a path from 
 * $s \in I(T)$ to $t \in I(T)$, where $s$ is the node where the predecessor of $v$ with the highest index was 
 * added to the decomposition for the first time. If the vertex $v$ has no predecessors, a path is built to an 
 * arbitrary node from the tree $T$. The bags on the resulting path between $s$ and $t$ contain only cliques.
 * <p>
 * The time complexity of this algorithm is $\mathcal{O}(|V|(|V|+|E|))$. In the decomposition produced by this 
 * class there are exactly $|V|-1$ many forget nodes since they are generated for every vertex in the $V(G)$ 
 * except for the vertex added as a root. There are at most $|V|-1$ additionally added join nodes which are 
 * generated during the connection of forget nodes to the tree. Every join node creates one additional 
 * root-to-leaf path. Therefore, there are $\mathcal{O}(|V|)$ root-to-leaf paths in the resulting decomposition. 
 * Every root-to-left path can contain $\mathcal{O}(|V|)$ introduce nodes, which yields $\mathcal{O}(|V|^2)$ 
 * upper bound on introduce nodes. Now considering the bags of the introduce nodes. Bags of every root-to-leaf 
 * path can contain at most $|V|$ distinct vertices of the initial graph. Every vertex $v$ can be introduced and 
 * forgotten on a path at most once. The sizes of these introduce and forget nodes are $\mathcal{O}(deg(v) + 1)$. 
 * Consequently, every root-to-leaf path is constructed in $\mathcal{O}(|V| + |E|)$ time. Since there are 
 * $\mathcal{O}(|V|)$ root-to-leaf paths, the total running time is $\mathcal{O}(|V|(|V| + |E|))$.
 * <p>
 * The algorithm used to construct a nice tree decomposition is a non-recursive adaption of algorithm 2 
 * from here: <br>
 * Hans L. Bodlaender, Arie M.C.A. Koster, Treewidth computations I. Upper bounds, Information and
 * Computation, Volume 208, Issue 3, 2010, Pages 259-275,
 * 
 * @author Ira Justus Fesefeldt (PhoenixIra)
 * @author Timofey Chudakov
 *
 * @since June 2018
 *
 * @param <V> the vertex type of the graph
 * @param <E> the edge type of the graph
 */
```
I've used your time analysis and mentioned a few relevant definitions.

Reply:
Thanks for the javadoc! I changed a small part regarding the forget nodes, which I came up with after you mentioned the join nodes: join nodes double existing nodes and adds a join node. Thus we can have 2(|V|-1) many forget nodes by doubling every forget node ones:

/**
 * Builds a nice tree decomposition of chordal graphs. See {@link NiceTreeDecompositionBase} for an
 * explanation of the nice tree decomposition.
 * <p>
 * For a given perfect elimination order of the graph $G = (V, E)$ every vertex in $V(G)$ has an
 * associated index in this order. The predecessors of a vertex $x \in V(G)$ are those vertices
 * incident to $x$ that have smaller index than $x$ (a vertex can have no predecessors). This class
 * uses a perfect elimination order produced by {@link ChordalityInspector} to iteratively add
 * vertices of $G$ to some nodes in the tree $T$. Initially, it adds a bag to the decomposition with
 * the first vertex from the perfect elimination order as a root of the tree. Then for all remaining
 * vertices $v \in V(G)$ this class generates a node (bag) $s \in I(T)$ in the tree $T$ which
 * contains vertex $v$ and all its predecessors. Then it builds a path from $s \in I(T)$ to $t \in
 * I(T)$, where $s$ is the node where the predecessor of $v$ with the highest index was added to the
 * decomposition for the first time. If the vertex $v$ has no predecessors, a path is built to an
 * arbitrary node from the tree $T$. The bags on the resulting path between $s$ and $t$ contain only
 * cliques.
 * <p>
 * The time complexity of this algorithm is $\mathcal{O}(|V|(|V|+|E|))$. In the decomposition
 * produced by this class there are at most $|V|-1$ additionally added join nodes which are
 * generated during the connection of forget nodes to the tree. The algorithm adds exactly $|V|-1$
 * many forget nodes since they are generated for every vertex in the $V(G)$ except for the vertex
 * added as a root. They may be doubled by join nodes, which adds up to at most $2(|V|-1)$ forget
 * nodes. Every join node creates one additional root-to-leaf path. Therefore, there are
 * $\mathcal{O}(|V|)$ root-to-leaf paths in the resulting decomposition. Every root-to-left path can
 * contain $\mathcal{O}(|V|)$ introduce nodes, which yields $\mathcal{O}(|V|^2)$ upper bound on
 * introduce nodes. Now considering the bags of the introduce nodes. Bags of every root-to-leaf path
 * can contain at most $|V|$ distinct vertices of the initial graph. Every vertex $v$ can be
 * introduced and forgotten on a path at most once. The sizes of these introduce and forget nodes
 * are $\mathcal{O}(deg(v) + 1)$. Consequently, every root-to-leaf path is constructed in
 * $\mathcal{O}(|V| + |E|)$ time. Since there are $\mathcal{O}(|V|)$ root-to-leaf paths, the total
 * running time is $\mathcal{O}(|V|(|V| + |E|))$.
 * <p>
 * The algorithm used to construct a nice tree decomposition is a non-recursive adaption of
 * algorithm 2 from here: <br>
 * Hans L. Bodlaender, Arie M.C.A. Koster, Treewidth computations I. Upper bounds, Information and
 * Computation, Volume 208, Issue 3, 2010, Pages 259-275,
 * 
 * @author Ira Justus Fesefeldt (PhoenixIra)
 * @author Timofey Chudakov
 *
 * @since June 2018
 *
 * @param <V> the vertex type of the graph
 * @param <E> the edge type of the graph
 */

Reply:
That’s right